cloudconfig/cloudinit/network_ubuntu.go

```diff
@@ -0,0 +1,432 @@
+	if len(interfaces) == 0 {
+		return "", errors.Errorf("missing container network config")
+	}
+	logger.Debugf("generating netplan from %#v", interfaces)
+	var netPlan netplan.Netplan
+	netPlan.Network.Ethernets = make(map[string]netplan.Ethernet)
+	netPlan.Network.Version = 2
+	for _, info := range interfaces {
+		var iface netplan.Ethernet
+		if cidr := info.CIDRAddress(); cidr != "" {
+			iface.Addresses = append(iface.Addresses, cidr)
+		} else if info.ConfigType == network.ConfigDHCP {
+			iface.Dhcp4 = true
+		}
+
+		for _, dns := range info.DNSServers {
+			iface.Nameservers.Addresses = append(iface.Nameservers.Addresses, dns.Value)
+		}
+		iface.Nameservers.Search = append(iface.Nameservers.Search, info.DNSSearchDomains...)
+
+		if info.GatewayAddress.Value != "" {
+			switch {
+			case info.GatewayAddress.Type == network.IPv4Address:
+				iface.Gateway4 = info.GatewayAddress.Value
+			case info.GatewayAddress.Type == network.IPv6Address:
+				iface.Gateway6 = info.GatewayAddress.Value
+			}
+		}
+
+		if info.MTU != 0 && info.MTU != 1500 {
+			iface.MTU = info.MTU
+		}
+		if info.MACAddress != "" {
+			iface.Match = map[string]string{"macaddress": info.MACAddress}
+		} else {
+			iface.Match = map[string]string{"name": info.InterfaceName}
+		}
+		for _, route := range info.Routes {
+			route := netplan.Route{
+				To:     route.DestinationCIDR,
+				Via:    route.GatewayIP,
+				Metric: route.Metric,
+			}
+			iface.Routes = append(iface.Routes, route)
+		}
+		netPlan.Network.Ethernets[info.InterfaceName] = iface
+	}
+	out, err := netplan.Marshal(netPlan)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return string(out), nil
+}
+
+// PreparedConfig holds all the necessary information to render a persistent
+// network config to a file.
+type PreparedConfig struct {
+	InterfaceNames   []string
+	AutoStarted      []string
+	DNSServers       []string
+	DNSSearchDomains []string
+	NameToAddress    map[string]string
+	NameToRoutes     map[string][]network.Route
+	NameToMTU        map[string]int
+	Gateway4Address  string
+	Gateway6Address  string
+}
+
+// PrepareNetworkConfigFromInterfaces collects the necessary information to
+// render a persistent network config from the given slice of
+// network.InterfaceInfo. The result always includes the loopback interface.
+func PrepareNetworkConfigFromInterfaces(interfaces []network.InterfaceInfo) *PreparedConfig {
+	dnsServers := set.NewStrings()
+	dnsSearchDomains := set.NewStrings()
+	gateway4Address := ""
+	gateway6Address := ""
+	namesInOrder := make([]string, 1, len(interfaces)+1)
+	nameToAddress := make(map[string]string)
+	nameToRoutes := make(map[string][]network.Route)
+	nameToMTU := make(map[string]int)
+
+	// Always include the loopback.
+	namesInOrder[0] = "lo"
+	autoStarted := set.NewStrings("lo")
+
+	// We need to check if we have a host-provided default GW and use it.
+	// Otherwise we'll use the first device with a gateway address,
```

Comment:
What if we have multiple devices with a gateway address? It seems possible we can set the wrong default gateway -- can we do better?